diff --git a/Makefile b/Makefile
index 9c7c7070f708c6adda780fbe6c200d5605054b7b..6f1944e22eb33331d9a586d2f9f6626418234f9c 100644
--- a/Makefile
+++ b/Makefile
@@ -13,9 +13,7 @@ CLEAN_FILES := $(TARGET) $(OBJS) *.disass *.sym *.map *.bin *.md5 *.hex *.elf *.
 
 rwildcard=$(foreach d,$(wildcard $(addsuffix *,$(1))),$(call rwildcard,$(d)/,$(2))$(filter $(subst *,%,$(2)),$(d)))
 
-CROSS_COMPILE_PATH := $(shell realpath $(CROSS_COMPILE)ld)
-CROSS_COMPILE_PATH := $(dir $(CROSS_COMPILE_PATH))
-$(info CROSS_COMPILE_PATH: $(CROSS_COMPILE_PATH))
+LIBGCC = $(shell $(CC) $(CPUTARGET) -print-libgcc-file-name)
 
 ifeq ($(SOC_TYPE), sky1)
 	SE_FW_SZ := 392703
@@ -25,16 +23,20 @@ else
 	SE_FW_SZ := 392703
 endif
 
-default: prebuild all
+default: all
 TARGET := se_fw.elf
-$(TARGET): build-subdirs $(OBJS) find-all-objs
-	@echo -e "\t" LD $(CFLAGS) $(ALL_OBJS) -L $(LDS_INC) -L $(CROSS_COMPILE_PATH)/../lib/gcc/arm-none-eabi/10.3.1/thumb/v8-m.main+fp/softfp/ -lgcc -T $(LDS) -o $@
-	@$(LD) $(ALL_OBJS) -L $(LDS_INC) -L $(CROSS_COMPILE_PATH)/../lib/gcc/arm-none-eabi/10.3.1/thumb/v8-m.main+fp/softfp/ -L $(CROSS_COMPILE_PATH)/platform/sky1 -L $(CROSS_COMPILE_PATH)/../../tools/gcc/gcc-arm-none-eabi-10.3-2021.10-x86_64-linux/bin/../lib/gcc/arm-none-eabi/10.3.1/thumb/v8-m.main+fp/softfp/ -lgcc -T $(LDS) -Map se_fw.map -o $@
+$(TARGET): $(OBJS) find-all-objs
+	@echo -e "\t" LD $(CFLAGS) $(ALL_OBJS) $(LIBGCC) -T $(LDS) -o $@
+	@$(LD) $(ALL_OBJS) -L $(LDS_INC) $(LIBGCC) -T $(LDS) -Map se_fw.map -o $@
 # @$(LD) $(ALL_OBJS) -L $(LDS_INC) -L $(CROSS_COMPILE_PATH)/../lib/gcc/arm-none-eabi/10.3.1/thumb/v8-m.main+fp/hard/ -lgcc -T $(LDS) -Map se_fw.map -o $@
 
 # phony targets
 .PHONY: all
-all: $(TARGET) postbuild
+all:
+	@$(MAKE) prebuild
+	@$(MAKE) build-subdirs
+	@$(MAKE) $(TARGET)
+	@$(MAKE) postbuild
 	#TODO, below are for make proxy library
 	#cd ./proxy && pwd && make lib_all
 	@echo Target $(TARGET) build finished.
@@ -66,11 +68,10 @@ postbuild:
 	$(OBJDUMP) -t $(TARGET) | sed '1,/SYMBOL TABLE/d; s/ .* / /; /^$$/d' > se_fw.sym
 	$(READELF) -A $(TARGET)
 	md5sum se_fw.elf > se_fw.md5
-	se_fw_sz_limit=$(SE_FW_SZ)
 	python3 ./tool/bin2hex/bin2hex.py  -bh se_fw.bin se_fw.hex
 	@file_size=$$(stat -c%s "se_fw.bin"); \
-	if [ $$file_size -gt $$se_fw_sz_limit ]; then \
-		echo "Error: se_fw.bin size ($$file_size bytes) exceeds 392703"; \
+	if [ $$file_size -gt $(SE_FW_SZ) ]; then \
+		echo "Error: se_fw.bin size ($$file_size bytes) exceeds $(SE_FW_SZ)"; \
 		rm se_fw.*; \
 		exit 1; \
 	else \
diff --git a/driver/dap/dap_rsp.c b/driver/dap/dap_rsp.c
index ffca5ff52c1b991a3043c123829de938d36503f7..01b48dceee78415ba24a4eb0563ec73ef06aa8ab 100644
--- a/driver/dap/dap_rsp.c
+++ b/driver/dap/dap_rsp.c
@@ -57,7 +57,7 @@ static uint32_t dap_packet_size = 64; //usb2.0
 
 static int rtp_rom_pwron(uint32_t base_address)
 {
-	uint val;
+	uint32_t val;
 
 	SMN_WRITEL(ARM_CS_SYSPCR_ROM_TABLE_PR, base_address + ARM_CS_ROM_TABLE_SYSPCR);
 	hal_mdelay(10);
@@ -376,7 +376,7 @@ static int rsp_query_cmd(char *data)
 		len = sprintf(data, "PacketSize=%08x#1", dap_buf.size);
 	} else if (!strncmp(data, "qSetPacketNum=", 14)) {
 		char *buf = data + 14;
-		uint num = strtoul(buf, NULL, 16);
+		uint32_t num = strtoul(buf, NULL, 16);
 		if (!num)
 			num = 1;
 		dap_buf.size = num * dap_packet_size;
diff --git a/driver/memory/drv_mem_config.c b/driver/memory/drv_mem_config.c
index 16e014666be67f80de3943d73beaf54ab79cc79d..f11c70ffd3184a7e565078d15159bf74e4a728d3 100644
--- a/driver/memory/drv_mem_config.c
+++ b/driver/memory/drv_mem_config.c
@@ -214,6 +214,8 @@ static errcode_t drv_memcfg_write_by_flash(uint32_t flash_offset, uint8_t *buf,
 		//LOG_I("flash write 4k success!\n");
 	} else {
 		LOG_I("flash write 4k fail, crc wrong!\n");
+		ret = FW_ERROR;
+		goto err_exit;
 	}
 
 	// judge if cross boundary
@@ -246,6 +248,8 @@ static errcode_t drv_memcfg_write_by_flash(uint32_t flash_offset, uint8_t *buf,
 			//LOG_I("flash write 4k success!\n");
 		} else {
 			LOG_I("flash write 4k fail, crc wrong!\n");
+			ret = FW_ERROR;
+			goto err_exit;
 		}
 	}
 
diff --git a/driver/uart/drv_uart.c b/driver/uart/drv_uart.c
index ae8255adad670d44cb11bf1cb1db7c3ba803065f..c8a7ad3852f99bcab291adb5e5d9c099803ebc09 100644
--- a/driver/uart/drv_uart.c
+++ b/driver/uart/drv_uart.c
@@ -187,7 +187,7 @@ errcode_t drv_putstring(const char *s, unsigned int count)
 /**
  * @brief uart getchar
  *
- * @param tmr_out uint in cpu cycles
+ * @param tmr_out uint32_t in cpu cycles
  *
  * @return the character it receives, otherwise, it is -1 indicates error.
  */
diff --git a/framework/fastboot/fw_fastboot.c b/framework/fastboot/fw_fastboot.c
index fb3e3e941ad77369c72ff834992bc4b5646ef3b8..aca5a76e6c92dd605b7400be8cf2c17376605d9c 100644
--- a/framework/fastboot/fw_fastboot.c
+++ b/framework/fastboot/fw_fastboot.c
@@ -137,7 +137,7 @@ static void _fw_fastboo_ack(const char *code, const char *info)
 	sprintf(gFastbootData.TxBuffer, "%s%s", code, info);
 
 	/** 3) Send TxBuffer to host side */
-	UsbDeviceControlSend(ENDPOINT_OUT, strlen(gFastbootData.TxBuffer), (uint64_t)(uint)gFastbootData.TxBuffer);
+	UsbDeviceControlSend(ENDPOINT_OUT, strlen(gFastbootData.TxBuffer), (uint64_t)(uint32_t)gFastbootData.TxBuffer);
 }
 
 static void _fw_fastboot_fail_ack(const char *Reason)
@@ -709,7 +709,7 @@ static void _fw_fastboot_process_rec_data(uint64_t Size, void *UsbDataBuffer)
 	} else {
 		/* Continue to receive data */
 		UsbDeviceControlSend(ENDPOINT_IN, _fw_fastboot_get_xfree_size(),
-				     (uint64_t)(uint)(UsbDataBuffer + gFastbootData.DataBytesReceived));
+				     (uint64_t)(uint32_t)(UsbDataBuffer + gFastbootData.DataBytesReceived));
 		hal_printf_level(LOG_LEVEL_CRIT, "AcceptData: Send %d\n", _fw_fastboot_get_xfree_size());
 	}
 #ifdef CONFIG_PERF_ENABLE
@@ -724,11 +724,11 @@ void _fw_fastboot_process_send_buf(void)
 	switch (gFastbootData.FastbootState) {
 	case CMD_STATE:
 		UsbDeviceControlSend(ENDPOINT_IN, _fw_fastboot_get_xfree_size(),
-				     (uint64_t)(uint)gFastbootData.RxBuffer);
+				     (uint64_t)(uint32_t)gFastbootData.RxBuffer);
 		break;
 	case DATA_STATE:
 		UsbDeviceControlSend(ENDPOINT_IN, _fw_fastboot_get_xfree_size(),
-				     (uint64_t)(uint)gFastbootData.UsbDataBuffer);
+				     (uint64_t)(uint32_t)gFastbootData.UsbDataBuffer);
 		break;
 	default:
 		hal_printf_level(LOG_LEVEL_CRIT, "Unknown fastboot state \r\n");
diff --git a/hal/crash/amu_einfo.c b/hal/crash/amu_einfo.c
index 74325a0cc86f22b21d43a7a8ed1dcaa005b29686..d70ffc750a6d1ba2edd74155ce4e10ca4b1b0aa1 100644
--- a/hal/crash/amu_einfo.c
+++ b/hal/crash/amu_einfo.c
@@ -7,24 +7,24 @@
 #define CPU_MAX_NUM (__builtin_ctzl(~FW_CORE_FULL_MASK))
 
 struct amu_data {
-	uint cpu;
-	uint cpu_cycle_low;
-	uint cpu_cycle_high;
-	uint cnt_cycle_low;
-	uint cnt_cycle_high;
-	uint cnt_inst_low;
-	uint cnt_inst_high;
-	uint memory_stall_cycle_low;
-	uint memory_stall_cycle_high;
+	uint32_t cpu;
+	uint32_t cpu_cycle_low;
+	uint32_t cpu_cycle_high;
+	uint32_t cnt_cycle_low;
+	uint32_t cnt_cycle_high;
+	uint32_t cnt_inst_low;
+	uint32_t cnt_inst_high;
+	uint32_t memory_stall_cycle_low;
+	uint32_t memory_stall_cycle_high;
 };
 
-void save_amu_data(struct amu_data *amu, uint cpu)
+void save_amu_data(struct amu_data *amu, uint32_t cpu)
 {
-	uint *data = &amu->cpu_cycle_low;
-	uint amu_base = AP_AMU_ADDR(cpu);
+	uint32_t *data = &amu->cpu_cycle_low;
+	uint32_t amu_base = AP_AMU_ADDR(cpu);
 
 	amu->cpu = cpu;
-	for (uint i = 0; i < 8; i++) {
+	for (uint32_t i = 0; i < 8; i++) {
 		data[i] = SMN_READL(amu_base + i * 4);
 	}
 }
@@ -33,13 +33,13 @@ void amu_write_einfo(void)
 {
 	struct se_mem_info *info = se_crash_open(SE_MEM_AMU);
 	struct amu_data d;
-	uint cpu, cpu_mask;
+	uint32_t cpu, cpu_mask;
 
 	if (!info)
 		return;
 
 	cpu_mask = fw_get_cpu_mask_by_harvest();
-	for (uint t = 0; t < 2; t++) {
+	for (uint32_t t = 0; t < 2; t++) {
 		se_crash_printf(info, "----------------time[%d]----------------\n", t);
 		for (cpu = 0; cpu < CPU_MAX_NUM; cpu++) {
 			if (!(BIT(cpu) & cpu_mask))
diff --git a/hal/crash/ddr_einfo.c b/hal/crash/ddr_einfo.c
index c8cbaa983d89c7d8afcc6332f19538f5c47a6dce..d014276ea26f5acec678bc1b6dc2680978d9017b 100644
--- a/hal/crash/ddr_einfo.c
+++ b/hal/crash/ddr_einfo.c
@@ -59,7 +59,7 @@ static const uint32_t ddr_rcsu_regs_bases[] = {
 
 bool ddr_check(void)
 {
-	uint full, tmp;
+	uint32_t full, tmp;
 
 	full = 0;
 	for (int i = 0; i < ARRAY_SIZE(ddr_rcsu_regs_bases); i++) {
@@ -70,23 +70,23 @@ bool ddr_check(void)
 	return full == 0;
 }
 
-static void ddr_save_channel(uint channel, void *addr)
+static void ddr_save_channel(uint32_t channel, void *addr)
 {
-	uint rcsu_base = ddr_rcsu_regs_bases[channel];
+	uint32_t rcsu_base = ddr_rcsu_regs_bases[channel];
 	struct ddr_channel_info *cinfo = addr;
 	struct ddr_reg *reg = &cinfo->regs[0];
-	uint index = 0;
+	uint32_t index = 0;
 
 	cinfo->channel = channel;
 	cinfo->num = REGS_NUM;
-	for (uint i = 0; i < ARRAY_SIZE(rcsu_regs_offsets); i++) {
+	for (uint32_t i = 0; i < ARRAY_SIZE(rcsu_regs_offsets); i++) {
 		reg[index].offset = rcsu_regs_offsets[i];
 		reg[index].type = TYPE_RCSU;
 		reg[index].val = SMN_READL(rcsu_base + rcsu_regs_offsets[i]);
 		index++;
 	}
 
-	for (uint i = 0; i < ARRAY_SIZE(ctrl_regs_offsets); i++) {
+	for (uint32_t i = 0; i < ARRAY_SIZE(ctrl_regs_offsets); i++) {
 		reg[index].offset = ctrl_regs_offsets[i];
 		reg[index].type = TYPE_CTRL;
 		reg[index].val = SMN_READL(rcsu_base + 0x10000 + ctrl_regs_offsets[i]);
@@ -97,7 +97,7 @@ static void ddr_save_channel(uint channel, void *addr)
 void ddr_einfo_save_to_sram(void)
 {
 	struct ddr_einfo *einfo_sram = (void *)RC_DDR_REG_TO_SRAM_BASE;
-	uint channel_start = 0, channel_num;
+	uint32_t channel_start = 0, channel_num;
 	uint32_t *crc = (void *)einfo_sram + EINFO_SIZE;
 
 	channel_num = drv_ddr_get_ch_num();
@@ -107,7 +107,7 @@ void ddr_einfo_save_to_sram(void)
 
 	einfo_sram->channel = channel_num;
 
-	for (uint i = channel_start; i < 4; i++) {
+	for (uint32_t i = channel_start; i < 4; i++) {
 		ddr_save_channel(i, &einfo_sram->info[i]);
 	}
 	*crc = crc32((void *)einfo_sram, EINFO_SIZE);
@@ -124,7 +124,7 @@ static bool ddr_sram_einfo_is_valid(struct ddr_einfo *einfo)
 	if (einfo->channel > 4 || !einfo->channel)
 		return false;
 
-	for (uint i = 0; i < einfo->channel; i++) {
+	for (uint32_t i = 0; i < einfo->channel; i++) {
 		if (einfo->info[i].num != REGS_NUM)
 			return false;
 	}
@@ -145,10 +145,10 @@ void ddr_einfo_copy_to_ddr(void)
 	if (!info)
 		return;
 
-	for (uint i = 0; i < einfo_sram->channel; i++) {
+	for (uint32_t i = 0; i < einfo_sram->channel; i++) {
 		cinfo = &einfo_sram->info[i];
 		se_crash_printf(info, "channel[%d]: num=%d\n", cinfo->channel, cinfo->num);
-		for (uint j = 0; j < cinfo->num; j++) {
+		for (uint32_t j = 0; j < cinfo->num; j++) {
 			se_crash_printf(info, "\toffset[%d]: 0x%x, type[%d]: 0x%x, val[%d]: 0x%x\n", j,
 					cinfo->regs[j].offset, j, cinfo->regs[j].type, j, cinfo->regs[j].val);
 		}
@@ -156,4 +156,4 @@ void ddr_einfo_copy_to_ddr(void)
 
 	se_crash_close(info);
 	memset(einfo_sram, 0, EINFO_SIZE + sizeof(uint32_t));
-}
\ No newline at end of file
+}
diff --git a/hal/crash/hal_crash.c b/hal/crash/hal_crash.c
index e6d97ad5f98041a8cab0f5151bb7312fdb0aa195..0651b42e46ac4787696c39b9a2558e45158e61a4 100644
--- a/hal/crash/hal_crash.c
+++ b/hal/crash/hal_crash.c
@@ -113,9 +113,9 @@ uint32_t crc32(uint8_t *data, uint32_t length)
 	return ~crc;
 }
 
-static uint hc_get_size(void)
+static uint32_t hc_get_size(void)
 {
-	uint size = 0;
+	uint32_t size = 0;
 
 	if (!g_rc || g_rc->flag == AP_USE)
 		goto out;
diff --git a/hal/crash/pm_wdt_einfo.c b/hal/crash/pm_wdt_einfo.c
index 8f49387d51c4ebf390f23143134ab261942ba839..932c1d03b49f4c5fb7ed2ed11abf0f4912ee4f0c 100644
--- a/hal/crash/pm_wdt_einfo.c
+++ b/hal/crash/pm_wdt_einfo.c
@@ -71,10 +71,10 @@ void se_save_pm_einfo(void)
 		return;
 
 	se_crash_printf(info, "   addr   ,    val    \n");
-	for (uint i = 0; i < REGS_NUM; i++)
+	for (uint32_t i = 0; i < REGS_NUM; i++)
 		se_crash_printf(info, "0x%08x: 0x%08x\n", pm_debug_regs[i], SMN_READL(pm_debug_regs[i]));
 
-	for (uint i = 0; i < sram_size; i++) {
+	for (uint32_t i = 0; i < sram_size; i++) {
 		uint32_t addr = RC_PM_DBG_SRAM_BASE + (i * sizeof(uint32_t));
 		se_crash_printf(info, "0x%08x: 0x%08x\n", addr, SMN_READL(addr));
 	}
diff --git a/platform/sky1/plat_monitor_clk.c b/platform/sky1/plat_monitor_clk.c
index c3c399f715b3a5f57fe3262268c6208657b784ea..1a3044d59b19ddcd8f674dada34d764c1a030632 100644
--- a/platform/sky1/plat_monitor_clk.c
+++ b/platform/sky1/plat_monitor_clk.c
@@ -1902,7 +1902,7 @@ void clk_save_einfo(void)
 	if (!info)
 		return;
 
-	for (uint i = 0; i < ARRAY_SIZE(clk_components); i++) {
+	for (uint32_t i = 0; i < ARRAY_SIZE(clk_components); i++) {
 		if (clk_components[i] == NULL)
 			break;
 
diff --git a/platform/sky1/plat_monitor_power.c b/platform/sky1/plat_monitor_power.c
index 2012ed0a69268a927678984d6cad5e775fa10e6e..09c220c0c86c1832ba6a8c89e77eba815cb1de90 100644
--- a/platform/sky1/plat_monitor_power.c
+++ b/platform/sky1/plat_monitor_power.c
@@ -297,7 +297,7 @@ void pwr_save_einfo(void)
 	if (!info)
 		return;
 
-	for (uint i = 0; i < ARRAY_SIZE(pwr_components); i++) {
+	for (uint32_t i = 0; i < ARRAY_SIZE(pwr_components); i++) {
 		if (pwr_components[i] == NULL)
 			break;
 
